app_dart/lib/src/request_handlers/dart_internal_subscription.dart

```diff
@@ -2,31 +2,208 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'package:cocoon_service/src/request_handling/body.dart';
+import 'package:cocoon_service/src/model/luci/buildbucket.dart';
+import 'package:gcloud/db.dart';
+import 'package:github/github.dart';
 import 'package:meta/meta.dart';
+import 'package:retry/retry.dart';
 
 import '../../cocoon_service.dart';
+import '../model/appengine/task.dart';
+import '../model/appengine/commit.dart';
+import '../request_handling/exceptions.dart';
 import '../request_handling/subscription_handler.dart';
+import '../service/datastore.dart';
+import '../service/exceptions.dart';
 import '../service/logging.dart';
 
+/// TODO(drewroengoogle): Make this subscription generic so we can accept more
+/// than just dart-internal builds.
+///
 /// An endpoint for listening to build updates for dart-internal builds and
 /// saving the results to the datastore.
 ///
 /// The PubSub subscription is set up here:
 /// https://console.cloud.google.com/cloudpubsub/subscription/detail/dart-internal-build-results-sub?project=flutter-dashboard
 @immutable
 class DartInternalSubscription extends SubscriptionHandler {
-  /// Creates an endpoint for listening to dart-internal build results.
+  /// Creates an endpoint for listening for dart-internal build results.
+  /// The message should contain a single buildbucket id
   const DartInternalSubscription({
     required super.cache,
     required super.config,
     super.authProvider,
+    required this.buildBucketClient,
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+    this.retryOptions = Config.buildbucketRetry,
   }) : super(subscriptionName: 'dart-internal-build-results-sub');
 
+  final BuildBucketClient buildBucketClient;
+  final DatastoreServiceProvider datastoreProvider;
+  final RetryOptions retryOptions;
+
   @override
   Future<Body> post() async {
-    log.fine('Consumed message: ${message.data}');
+    final DatastoreService datastore = datastoreProvider(config.db);
+
+    log.info("Converting message data to non nullable int");
+
+    final int buildbucketId = int.parse(message.data.toString().replaceAll("'", ""));
+
+    log.info("Creating build request object");
+    final GetBuildRequest request = GetBuildRequest(id: buildbucketId.toString());
+
+    log.info(
+      "Calling buildbucket api to get build data for build $buildbucketId",
+    );
+    late Build build;
+    try {
+      build = await retryOptions.retry(
+        () async {
+          return _getBuildFromBuildbucket(request);
+        },
+        retryIf: (Exception e) => e is UnfinishedBuildException,
+      );
+    } catch (e) {
+      if (e is! UnfinishedBuildException) {
```

Comment:
Why do we need this if?

Reply:
My intention with this is to be able to try getting the build a couple times in case it hasn't finished. However, in the case that the build is still not complete, but the build doesn't represent that, I didn't want to fail the build.

Retry does not support retrying based on the returned object, would it make sense to add a custom retry instead?

Reply:
I've updated it to fail if we are unable to get a complete build after three attempts.